File: algorithms/2_medium/M/00542_01_matrix/bfs.py
DIRECTIONS = [(1, 0), (-1, 0), (0, 1), (0, -1)]

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        rows = len(mat)
        cols = len(mat[0])

        res = [[0] * cols for _ in range(rows)]
        is_visited = [[False] * cols for _ in range(rows)]
        q = deque()

        for i in range(rows):
            for j in range(cols):
                if mat[i][j] == 0:
                    q.append((i, j))
                    is_visited[i][j] = True

        curr_level = 0

        while q:
            curr_len = len(q)

            for _ in range(curr_len):
                (i, j) = q.popleft()
                res[i][j] = curr_level

                for (dx, dy) in DIRECTIONS:
                    new_i = i + dx
                    new_j = j + dy

                    if new_i >= rows or new_i < 0 or new_j >= cols or new_j < 0 or is_visited[new_i][new_j]:
                        continue

                    is_visited[new_i][new_j] = True
                    q.append((new_i, new_j))

            curr_level += 1

        return res

File: algorithms/2_medium/M/00542_01_matrix/bfs.rs
use std::collections::VecDeque;

impl Solution {
    const DIRECTIONS: [(isize, isize); 4] = [(1, 0), (-1, 0), (0, 1), (0, -1)];

    pub fn update_matrix(mat: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let rows = mat.len();
        let cols = mat[0].len();

        let mut res = vec![vec![0_i32; cols]; rows];
        let mut is_visited = vec![vec![false; cols]; rows];
        let mut q: VecDeque<(usize, usize)> = VecDeque::new();

        for i in 0..rows {
            for j in 0..cols {
                if mat[i][j] == 0 {
                    q.push_back((i, j));
                    is_visited[i][j] = true;
                }
            }
        }

        let mut curr_level = 0_i32;

        while !q.is_empty() {
            let curr_len = q.len();

            for _ in 0..curr_len {
                let (i, j) = q.pop_front().unwrap();
                res[i][j] = curr_level;

                for (dx, dy) in Self::DIRECTIONS.iter() {
                    let new_i = (i as isize + dx) as usize;
                    let new_j = (j as isize + dy) as usize;

                    if new_i >= rows || new_j >= cols || is_visited[new_i][new_j] {
                        continue;
                    }

                    is_visited[new_i][new_j] = true;
                    q.push_back((new_i, new_j));
                }
            }

            curr_level += 1;
        }

        res
    }
}
